src/timer.h

```diff
@@ -1,49 +1,54 @@
 /*
  * This file is
  * Copyright (c) 2009 by Jim Fougeron jfoug AT cox dot net
- * Copyright (c) 2019 by magnum
+ * Copyright (c) 2019-2020 by magnum
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted.
- *
- * Portable hi-res timer.  Was a nice C++ class. Downgraded to C for john.
  */
 
 #ifndef _HAVE_TIMER_H
 #define _HAVE_TIMER_H
 
-#include <time.h>
-
 /* Windows - use provided hi-res timer */
 #if defined (_MSC_VER) || defined (__MINGW32__) || defined (__CYGWIN32__)
 
-#include <sys/timeb.h>      /* for ftime(), which is not used */
 #undef MEM_FREE
 #include <windows.h>
 #undef MEM_FREE
+
+/*
+ * QueryPerformanceCounter(LARGE_INTEGER*) gives "counts" (ticks)
+ * QueryPerformanceFrequency(LARGE_INTEGER*) gives ticks/s
+ */
+#define HR_CLOCK          "QueryPerformanceCounter()"
 typedef LARGE_INTEGER hr_timer;
 
-#define HRZERO(X)             (X).HighPart = (X).LowPart = 0
-#define HRSETCURRENT(X)       QueryPerformanceCounter(&(X))
-#define HRGETTICKS(X)         ((uint64_t)(X).HighPart * 4294967296ULL + \
-                               (uint64_t)(X).LowPart)
-#define HRGETTICKS_PER_SEC(X) {	  \
-		LARGE_INTEGER large; \
-		if (QueryPerformanceFrequency (&large)) \
-			(X) = (uint64_t)large.HighPart*4294967296ULL + (uint64_t)large.LowPart; \
-		else \
-			(X) = 0; \
-	}
+#define HRZERO(X)         (X).QuadPart = 0
+#define HRSETCURRENT(X)   QueryPerformanceCounter(&(X))
+#define HRGETTICKS(X)     ((X).QuadPart)
+#define HRSETRESOLUTION() ({ LARGE_INTEGER t; QueryPerformanceFrequency(&t); timer_ticks_per_sec = t.QuadPart; })
+#define HRTICKS_PER_SEC   (timer_ticks_per_sec)
+#define timer_get_nano()  ({ hr_timer t; HRSETCURRENT(t); t.QuadPart * 1000000000ULL / timer_ticks_per_sec; })
```

Comment:
You're now heavily using GCC's "Statement Exprs" extension in `timer.h`, including in its parts that are obviously meant to also be compiled by non-GCC. I expect this will break at least with MSVC. I'm afraid you'll need to reconsider that decision throughout `timer.h`.

Reply:
@crioux was also builting a lib'ified version of jumbo with MSVC for LC7, but I don't know if any further updates of that are planned (perhaps not).

Anyway, this isn't only about MSVC. So far, we did not require GCC specifics, and I think it'd be wrong for us to change that just for the revised hi-res timer code. There's more value in not depending on GCC than in the new timer code.

Reply:
- [ ] [google.com]()CodeQL analysis
@vanessayuenn
Hello git history spelunker!
- 3 months33:"")ago
2
 
push:
@rneatherway
@heiskr
Update codeql.yml ('m16791)
22 days ago
5
    branches: main
6
  pull_request:
7
    branches: main
@vanessayuenn
Hello git history spelunker!
3 months ago
8
    paths:
@nschonni
chore: Add Prettier for Yaml formatting
2 months ago
9
      - '**/*.js'
10
      - '.github/workflows/codeql.yml'
@vanessayuenn
Hello git history spelunker!
3 months ago
11

12
jobs:
13
  build:
@nschonni
chore: Add Prettier for Yaml formatting
2 months ago
14
    runs-on: ubuntu-latest
@vanessayuenn
Hello git history spelunker!
3 months ago
15

16
    steps:
@nschonni
chore: Add Prettier for Yaml formatting
2 months ago
17
      - uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f
18
      - uses: github/codeql-action/init@v1
19
        with:
20
          languages: javascript # comma separated list of values from {go, python, javascript, java, cpp, csharp} (not YET ruby, sorry!)
21
      - uses: github/codeql-action/analyze@v1
22
        continue-on-error: true